Use bitcoin kmp

sstone opened the pull request:
There is one breaking change: I've removed bouncycastle, so most of bitcoin-lib cannot be used if secp256k1-kmp cannot be loaded. Which means that we are now restricted to Windows 64 bits, Macos 64 bits, Linux 64 bits, iOS, Android.
There may be issues with strange/custom linux distros but I'm not aware of them yet.

sstone commented on src/test/scala/fr/acinq/bitcoinscala/DeterministicWalletSpec.scala:
```diff
@@ -114,25 +114,6 @@ class DeterministicWalletSpec extends FlatSpec {
     assert(encode(publicKey(m_0h_1h), xpub) === "xpub6BJA1jSqiukeaesWfxe6sNK9CCGaujFFSJLomWHprUL9DePQ4JDkM5d88n49sMGJxrhpjazuXYWdMf17C9T5XnxkopaeS7jGk1GyyVziaMt")
   }
 
-  it should "be possible to go up the private key chain if you have the master pub key and a child private key!!" in {
```
done in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/7767c9ace6eba6f81d92adb6ad573a6191b26da2

sstone commented on src/main/scala/fr/acinq/bitcoinscala/KotlinUtils.scala:
```diff
@@ -0,0 +1,191 @@
+import scala.jdk.CollectionConverters.{ListHasAsScala, SeqHasAsJava}
+
+object KotlinUtils {
+  implicit def kmp2scala(input: fr.acinq.bitcoin.ByteVector32): ByteVector32 = ByteVector32(ByteVector(input.toByteArray))
```
done in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/7767c9ace6eba6f81d92adb6ad573a6191b26da2

sstone commented on src/main/scala/fr/acinq/bitcoinscala/KotlinUtils.scala:
```diff
@@ -0,0 +1,191 @@
+  implicit def scala2kmp(input: PublicKey): bitcoin.PublicKey = new bitcoin.PublicKey(input.value)
+
+  case class InputStreamWrapper(is: InputStream) extends fr.acinq.bitcoin.io.Input {
+    override def getAvailableBytes: Int = is.available()
```
On the JVM we use `ByteArrayInputStream`, its implementation of `available()`  provides the number of bytes that are actually available

sstone commented on src/main/scala/fr/acinq/bitcoinscala/Script.scala:
```diff
@@ -0,0 +1,411 @@
+
+  def isDisabled(op: ScriptElt) = fr.acinq.bitcoin.Script.isDisabled(op)
+
+  def cost(op: ScriptElt): Int = fr.acinq.bitcoin.Script.cost(op)
```
removed in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/7767c9ace6eba6f81d92adb6ad573a6191b26da2

sstone commented on src/main/scala/fr/acinq/bitcoinscala/Script.scala:
```diff
@@ -0,0 +1,411 @@
+    case _ => false
+  }
+
+  def isNativeWitnessScript(script: ByteVector): Boolean = isNativeWitnessScript(parse(script))
```
done  in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/7767c9ace6eba6f81d92adb6ad573a6191b26da2

t-bast commented:
I fixed most of my comments in this commit if you want to cherry-pick it: https://github.com/ACINQ/bitcoin-lib/commit/350b2da1b8605931cee9580c8ab9f5bb80665c4f

Note however that it needs an update to the latest version of `bitcoin-kmp` because it calls some functions that were added recently.

sstone commented:
Yes I've released https://github.com/ACINQ/secp256k1-kmp/releases/tag/v0.6.1, I'll release `bitcoin-kmp 0.8.1` asap and then update this PR

sstone commented:
I've updated `bitcoin-kmp` and `secp256k1-kmp`, cherry-picked https://github.com/ACINQ/bitcoin-lib/commit/350b2da1b8605931cee9580c8ab9f5bb80665c4f and updated https://github.com/ACINQ/eclair/pull/2038

t-bast commented:
I think we agree that this is going in the right direction, so I believe we should quickly integrate this PR and then potentially do follow-up PRs for additional clean-up and add more tests.

It's quite hard to review because github marks mostly everything as new files, so let's merge this soon and work on top before we actually make a release?

The only thing that's blocking my `ACK` is doing performance tests to have an overall idea of how it performs compared to the full scala version.

sstone commented on src/main/scala/fr/acinq/bitcoinscala/DeterministicWallet.scala:
```diff
@@ -0,0 +1,186 @@
+
+  def encode(input: ExtendedPublicKey, prefix: Int): String = bitcoin.DeterministicWallet.encode(input.pub, prefix)
+
+  def write(input: ExtendedPublicKey, output: OutputStream): Unit = {
+    writeUInt8(input.depth, output)
+    writeUInt32(input.parent.toInt, output, ByteOrder.BIG_ENDIAN)
+    writeUInt32(input.path.lastChildNumber.toInt, output, ByteOrder.BIG_ENDIAN)
+    writeBytes(input.chaincode.toArray, output)
+    writeBytes(input.publickeybytes.toArray, output)
+  }
```
I think we can just remove this one, it's probably a remnant of when we tried to use java serialization.

sstone replied:
done in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/2768905c8ef713ae87a2d5fba9bfd032026339af

sstone commented on src/main/scala/fr/acinq/bitcoinscala/MerkleTree.scala:
```diff
@@ -0,0 +1,12 @@
+  * see https://en.bitcoin.it/wiki/Protocol_specification#Merkle_Trees
+  */
+object MerkleTree {
+  def computeRoot(tree: Seq[ByteVector32]): ByteVector32 = fr.acinq.bitcoin.MerkleTree.INSTANCE.computeRoot(tree.map(scala2kmp).asJava)
```
Yes it's correct.

sstone commented on src/main/scala/fr/acinq/bitcoinscala/Crypto.scala:
```diff
@@ -0,0 +1,226 @@
+
+    def add(that: PublicKey): PublicKey = PublicKey(this.pub plus that.pub)
+
+    def add(that: PrivateKey): PublicKey = PublicKey(this.pub plus that.priv.publicKey())
```
done in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/2768905c8ef713ae87a2d5fba9bfd032026339af

sstone commented on src/main/scala/fr/acinq/bitcoinscala/Crypto.scala:
```diff
@@ -0,0 +1,226 @@
+
+    def add(that: PrivateKey): PublicKey = PublicKey(this.pub plus that.priv.publicKey())
+
+    def subtract(that: PublicKey): PublicKey = PublicKey(this.pub plus that.pub)
```
done in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/2768905c8ef713ae87a2d5fba9bfd032026339af, I added a few simple tsets

sstone commented on src/main/scala/fr/acinq/bitcoinscala/package.scala:
```diff
@@ -0,0 +1,100 @@
+   * @param script    public key script
+   * @return the address of this public key script on this chain
+   */
+  def computeScriptAddress(chainHash: ByteVector32, script: Seq[ScriptElt]): String = {
```
done in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/2768905c8ef713ae87a2d5fba9bfd032026339af

sstone commented:
https://github.com/ACINQ/eclair-benchmarks/tree/master/eclair-benchmark does not show differences between eclair on master and eclair on https://github.com/ACINQ/eclair/pull/2038:

master:
```
# Benchmark: fr.acinq.eclair.CommitmentsBenchmark.testPayment

# Run progress: 0.00% complete, ETA 00:16:40
# Fork: 1 of 5
# Warmup Iteration   1: setup
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
SLF4J: Failed to load class "org.slf4j.impl.StaticMDCBinder".
SLF4J: Defaulting to no-operation MDCAdapter implementation.
SLF4J: See http://www.slf4j.org/codes.html#no_static_mdc_binder for further details.
42.058 ops/s
# Warmup Iteration   2: 44.708 ops/s
# Warmup Iteration   3: 45.197 ops/s
# Warmup Iteration   4: 45.287 ops/s
# Warmup Iteration   5: 45.398 ops/s
Iteration   1: 45.377 ops/s
Iteration   2: 45.814 ops/s
Iteration   3: 45.428 ops/s
Iteration   4: 45.368 ops/s
Iteration   5: 45.339 ops/s
```

https://github.com/ACINQ/eclair/pull/2038:
```
# Benchmark: fr.acinq.eclair.CommitmentsBenchmark.testPayment

# Run progress: 0.00% complete, ETA 00:16:40
# Fork: 1 of 5
# Warmup Iteration   1: setup
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
SLF4J: Failed to load class "org.slf4j.impl.StaticMDCBinder".
SLF4J: Defaulting to no-operation MDCAdapter implementation.
SLF4J: See http://www.slf4j.org/codes.html#no_static_mdc_binder for further details.
42.037 ops/s
# Warmup Iteration   2: 45.307 ops/s
# Warmup Iteration   3: 46.151 ops/s
# Warmup Iteration   4: 46.322 ops/s
# Warmup Iteration   5: 46.391 ops/s
Iteration   1: 46.378 ops/s
Iteration   2: 46.441 ops/s
Iteration   3: 46.393 ops/s
Iteration   4: 46.344 ops/s
Iteration   5: 46.411 ops/s
``` 

I still would like to check if/how GC is affected, I'll comment on that here

t-bast reviewed (commented):
LGTM, let's just wait for GC impact results.

sstone commented:
I've updated the jmh benchmarks to get actors out of the way (creating probes took most of the times) and now we get:

With bitcoin-lib ([eclair on `master`](https://github.com/ACINQ/eclair)):
```
Result "fr.acinq.eclair.CommitmentsBenchmark.sendCommit":
  340.141 ±(99.9%) 1.256 ops/s [Average]
  (min, avg, max) = (337.756, 340.141, 343.078), stdev = 1.677
  CI (99.9%): [338.885, 341.398] (assumes normal distribution)
```

With bitcoin-kmp ([eclair on `use-bitcoin-kmp-through-bitcoin-lib`](https://github.com/ACINQ/eclair/tree/use-bitcoin-kmp-through-bitcoin-lib))
 
``` 
 Result "fr.acinq.eclair.CommitmentsBenchmark.sendCommit":
  355.298 ±(99.9%) 1.018 ops/s [Average]
  (min, avg, max) = (353.025, 355.298, 357.577), stdev = 1.358
  CI (99.9%): [354.281, 356.316] (assumes normal distribution)
```

There are no significant differences in speed or memory usage, and it's consistent with lightning-kmp benchmarks

t-bast commented on src/main/scala/fr/acinq/bitcoinscala/Crypto.scala:
```diff
@@ -0,0 +1,226 @@
+  }
+
+  object PrivateKey {
+    def apply(data: ByteVector): PrivateKey = new PrivateKey(new bitcoin.PrivateKey(data.toArray))
```
```suggestion
    def apply(data: ByteVector): PrivateKey = PrivateKey(new bitcoin.PrivateKey(data.toArray))
```

t-bast commented on src/main/scala/fr/acinq/bitcoinscala/DeterministicWallet.scala:
```diff
@@ -0,0 +1,186 @@
+  object KeyPath {
+    val Root = KeyPath(new bitcoin.KeyPath(""))
+
+    def apply(path: Seq[Long]): KeyPath = new KeyPath(new bitcoin.KeyPath(path.map(x => Long.box(x)).toList.asJava))
```
```suggestion
    def apply(path: Seq[Long]): KeyPath = KeyPath(new bitcoin.KeyPath(path.map(x => Long.box(x)).toList.asJava))
```

t-bast commented on src/main/scala/fr/acinq/bitcoinscala/DeterministicWallet.scala:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  object ExtendedPrivateKey {
+    def apply(secretkeybytes: ByteVector32, chaincode: ByteVector32, depth: Int, derivationPath: KeyPath, parent: Long) = new ExtendedPrivateKey(
```
```suggestion
    def apply(secretkeybytes: ByteVector32, chaincode: ByteVector32, depth: Int, derivationPath: KeyPath, parent: Long) = ExtendedPrivateKey(
```

t-bast commented on src/main/scala/fr/acinq/bitcoinscala/DeterministicWallet.scala:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  object ExtendedPublicKey {
+    def apply(publicKey: ByteVector, chaincode: ByteVector32, depth: Int, derivationPath: KeyPath, parent: Long) = new ExtendedPublicKey(
```
```suggestion
    def apply(publicKey: ByteVector, chaincode: ByteVector32, depth: Int, derivationPath: KeyPath, parent: Long) = ExtendedPublicKey(
```

t-bast reviewed (commented):
LGTM, a few nits about `case class` instantiation.
I'll review https://github.com/ACINQ/eclair/pull/2038 and if it's good to go, we can pull the plug!

pm47 reviewed (changes requested):
> I've removed bouncycastle [...] Which means that we are now restricted to Windows 64 bits, Macos 64 bits, Linux 64 bits, iOS, Android.

I'm fine with dropping bouncycastle, but we must support arm64 platform.

pm47 commented:
I would rename `fr.acinq.bitcoinscala` to `fr.acinq.bitcoin.scalacompat` for the following reasons:
1) only one package hierarchy `fr.acinq.bitcoin.*`
2) the meaning of `fr.acinq.bitcoin.scalacompat` is self explanatory

sstone commented:
> > I've removed bouncycastle [...] Which means that we are now restricted to Windows 64 bits, Macos 64 bits, Linux 64 bits, iOS, Android.
> 
> I'm fine with dropping bouncycastle, but we must support arm64 platform.

I tested that the instructions that we document in https://github.com/ACINQ/secp256k1-kmp/blob/master/README.md#jvm-targets--android work on linux ARM64 (i.e. it's very easy to compile our JNI wrapper and use it from eclair).

sstone commented:
> I'm fine with dropping bouncycastle, but we must support arm64 platform.

Done in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/55a91b57bd0927c966b6ed59b3550a7c288c4f82: https://github.com/ACINQ/secp256k1-kmp (which we get through https://github.com/ACINQ/bitcoin-kmp) includes JNI bindings for Linux Arm64 from version 0.6.3 on.

t-bast commented on src/main/scala/fr/acinq/bitcoin/scalacompat/Crypto.scala:
```diff
@@ -0,0 +1,226 @@
+
+    def add(that: PrivateKey): PublicKey = PublicKey(this.pub plus that.priv.publicKey())
+
+    def subtract(that: PublicKey): PublicKey = PublicKey(this.pub plus that.pub)
```
It's weird, I think we had previously fixed that in earlier commits?

```suggestion
    def subtract(that: PublicKey): PublicKey = PublicKey(this.pub minus that.pub)
```

sstone replied:
I thought so too, fixed in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/c67bb611f713af65c71ebea6354db1b9ca951a91

t-bast commented on src/test/scala/fr/acinq/bitcoin/scalacompat/reference/TransactionSpec.scala:
```diff
@@ -65,12 +67,12 @@ class TransactionSpec extends FlatSpec with Matchers {
 
   "Bitcoins library" should "pass reference tx valid tests" in {
-    val stream = classOf[ScriptSpec].getResourceAsStream("/data/tx_valid.json")
+    val stream = classOf[scalacompat.ScriptSpec].getResourceAsStream("/data/tx_valid.json")
```
Shouldn't this be `TransactionSpec` (and same below)?

sstone replied:
Yes it would be cleaner, done in https://github.com/ACINQ/bitcoin-lib/pull/63/commits/c67bb611f713af65c71ebea6354db1b9ca951a91
